assets/src/edit-story/components/panels/size.js

```diff
@@ -18,101 +18,216 @@
  * External dependencies
  */
 import PropTypes from 'prop-types';
+import styled from 'styled-components';
 
 /**
  * WordPress dependencies
  */
-import { useEffect, useState } from '@wordpress/element';
+import { useEffect, useState, useCallback } from '@wordpress/element';
 import { __, _x } from '@wordpress/i18n';
 
 /**
  * Internal dependencies
  */
-import { InputGroup } from '../form';
+import { Spacer, Button, Row, Numeric } from '../form';
 import { dataPixels } from '../../units';
+import Locked from '../../icons/lock.svg';
+import Unlocked from '../../icons/unlock.svg';
+import Fullbleed from '../../icons/fullbleed.svg';
+import FlipVertical from '../../icons/flip_vertical.svg';
+import FlipHorizontal from '../../icons/flip_horizontal.svg';
+import Toggle from '../form/toggle';
 import { SimplePanel } from './panel';
 import getCommonValue from './utils/getCommonValue';
 
+const ExpandedNumeric = styled(Numeric)`
+  flex: 1;
+`;
+
+const BoxedNumeric = styled(ExpandedNumeric)`
+  padding: 6px 6px;
+  border: 1px solid ${({ theme }) => theme.colors.fg.v3};
+  border-radius: 4px;
+`;
+
 function SizePanel({ selectedElements, onSetProperties }) {
+  const x = getCommonValue(selectedElements, 'x');
+  const y = getCommonValue(selectedElements, 'y');
   const width = getCommonValue(selectedElements, 'width');
   const height = getCommonValue(selectedElements, 'height');
   const isFill = getCommonValue(selectedElements, 'isFill');
-  const [state, setState] = useState({ width, height });
+  const rotationAngle = getCommonValue(selectedElements, 'rotationAngle');
+  const [state, setState] = useState({
+    x,
+    y,
+    width,
+    height,
+    isFill,
+    rotationAngle,
+  });
   const [lockRatio, setLockRatio] = useState(true);
   useEffect(() => {
-    setState({ width, height });
-  }, [width, height]);
-  const handleSubmit = (evt) => {
-    onSetProperties(({ width: oldWidth, height: oldHeight }) => {
-      let { width: newWidth, height: newHeight } = state;
-      if (lockRatio && (newHeight === '' || newWidth === '')) {
-        const ratio = oldWidth / oldHeight;
-        if (newWidth === '') {
-          newWidth = dataPixels(newHeight * ratio);
-        } else {
-          newHeight = dataPixels(newWidth / ratio);
+    setState({ x, y, width, height, isFill, rotationAngle });
+  }, [x, y, width, height, isFill, rotationAngle]);
+  useEffect(() => {
+    updateProperties();
+  }, [state.isFill, updateProperties]);
+  const updateProperties = useCallback(
+    (evt) => {
+      onSetProperties(({ width: oldWidth, height: oldHeight }) => {
+        const { height: newHeight, width: newWidth, ...rest } = state;
+        const update = { width: newWidth, height: newHeight, ...rest };
+        if (
+          lockRatio &&
+          (newHeight === '' || newWidth === '') &&
+          !(newHeight === '' && newWidth === '')
+        ) {
```

Comment:
This should be written with some `const`'s above to explain what's going on here. Something like `if (lockRatio && (hasHeightButNotWidth || hasWidthButNotHeight))` or preferably something even better than that 😉

Reply:
Cool. I will address this.